File: PHDtools/html/Diagnosis-tools/conserve-seq.php
<?php
    $visualization = "visibility:visible";
    $dtime = date('Y-m-d-h-i-s', time());
    $perl_path = "/home/dell/miniconda3/envs/Perl/bin/perl";
    $analysis_path = "/home/dell/data/Web-Server/ConserveSeq/".$dtime."/";
    $cons_pipeline = "/var/www/other_scripts/bio-pipelines/Conserve-sequence/conserve-seq.pl";
    $Result_store_path = "/home/dell/data/Web-Server/Result-store/ConserveSeq/".$dtime."/";
    if($_FILES['genomes']['error'] == 0 && $_FILES['reference']['error'] == 0){
        if($_FILES['genome']['size'] > 104857600){
            die('The upload genome sequences file should be smaller than 100Mb!<br>');
        }
        $iso_filename = $_FILES['genomes']['name'];
        $Ref_filename = $_FILES['reference']['name'];
        $iso_tmp = $_FILES['genomes']['tmp_name'];
        $Ref_tmp = $_FILES['reference']['tmp_name'];
	shell_exec("mkdir ".$analysis_path);
	shell_exec("mkdir ".$analysis_path.".../");
        shell_exec("chmod 777 -R ".$analysis_path);
        if(move_uploaded_file($iso_tmp, $analysis_path.$iso_filename)){
            echo "$iso_filename isolate genomes file uploaded!<br>";
            shell_exec("chmod 777 -R ".$analysis_path);
            if($_FILES['genomes']['type'] == "application/octet-stream"){
                shell_exec("mv ".$analysis_path.$iso_filename." ".$analysis_path."isolate_genomes.fasta");
                $iso_filename = "isolate_genomes.fasta";
                shell_exec("chmod -R 777 ".$analysis_path);
            }else if($_FILES['genomes']['type'] == "application/gzip"){
                shell_exec("zcat ".$analysis_path.$iso_filename." > ".$analysis_path."isolate_genomes.fasta");
                $iso_filename = "isolate_genomes.fasta";
                shell_exec("chmod -R 777 ".$analysis_path);
            }else if($_FILES['genomes']['type'] == "application/zip" || $_FILES['genomes']['type'] == "application/x-zip-compressed"){
                shell_exec("unzip -o ".$analysis_path.$iso_filename." -d ".$analysis_path."isolate_genomes.fasta");
                $iso_filename = "isolate_genomes.fasta";
                shell_exec("chmod -R 777 ".$analysis_path);
            }else{
                shell_exec("rm -rf ".$analysis_path);
                die('The pathogen genomes shoud be .fasta/.fasta.gz/.fasta.zip format<br>');
            }
        }else{
            shell_exec("rm -rf ".$analysis_path);
            die('$iso_filename isolate genomes file upload filed!<br>');
        }
        if(move_uploaded_file($Ref_tmp, $analysis_path.$Ref_filename)){
            echo "$Ref_filename Reference sequence uploaded!<br>";
            shell_exec("chmod 777 -R ".$analysis_path);
            if($_FILES['reference']['type'] == "application/octet-stream"){
                shell_exec("mv ".$analysis_path.$Ref_filename." ".$analysis_path."Ref_genome.fasta");
                $Ref_filename = "Ref_genome.fasta";
                shell_exec("chmod -R 777 ".$analysis_path);
            }else if($_FILES['reference']['type'] == "application/gzip"){
                shell_exec("zcat ".$analysis_path.$Ref_filename." > ".$analysis_path."Ref_genome.fasta");
                shell_exec("rm ".$analysis_path.$Ref_filename);
                $Ref_filename = "Ref_genome.fasta";
                shell_exec("chmod -R 777 ".$analysis_path);
            }else if($_FILES['reference']['type'] == "application/zip" || $_FILES['reference']['type'] == "application/x-zip-compressed"){
                shell_exec("unzip -o ".$analysis_path.$Ref_filename." -d ".$analysis_path."Ref_genome.fasta");
                shell_exec("rm ".$analysis_path.$Ref_filename);
                $Ref_filename = "Ref_genome.fasta";
                shell_exec("chmod -R 777 ".$analysis_path);
            }else{
                shell_exec("rm -rf ".$analysis_path);
                die('The reference genome shoud be .fasta/.fasta.gz/.fasta.zip format<br>');
            }
        }else{
            shell_exec("rm -rf ".$analysis_path);
            die('$Ref_filename Reference sequence upload filed!<br>');
        }
        if(is_numeric($_POST['targetLen'])){
            $targetLen = $_POST['targetLen'];
            if($targetLen > 1000){
                shell_exec("rm -rf ".$analysis_path);
                die('The expected target length of conserved sequence must be shorter than 1000<br>');
            }
        }else{
            shell_exec("rm -rf ".$analysis_path);
            die('The input of the expected target length must be a numeric value!<br>');
		}
        $command = $perl_path." ".$cons_pipeline." -WorkPath ".$analysis_path." -reference ".$Ref_filename." -isolates ".$iso_filename." -algo ".$_POST['algo']." -targetLen ".$targetLen." -StorePath ".$Result_store_path;
        shell_exec($command);
        shell_exec("chmod 777 ".$analysis_path.".../shell.sh");
        if($_POST['waiting'] == "simultaneous"){
            $condition = "TRUE";
            shell_exec("bash ".$analysis_path.".../shell.sh");
            echo "The analysis finished, the results can be download from the follows link:<br>";
            $download_path = "../Download-tools/doDownload.php?filename=".$Result_store_path."MostConserve_info.zip";
        }else{
            $visualization = "visibility:hidden";
            echo "Your Download code is: ConserveSeq_".$dtime."<br>";
            echo "Please waiting for the analysis, remember the download code and you can close the window!<br>";
            pclose(popen('bash '.$analysis_path.'.../shell.sh &', 'r'));
        }
    }else{
        die('Files upload error!');
    }
?>

<html lang="en">
    <head>
        <title>The most conserved sequence identification</title>
            <meta charset="UTF-8"/>
    </head>
<body>
    <a href="<?php echo $download_path?>" style="<?php echo $visualization?>">DownloadConserveInfoFile</a><br>
</body>
</html>

File: PHDtools/other_scripts/bio-pipelines/Conserve-sequence/conserve-seq.pl
#!/usr/bin/perl -w

=head1 Description
	Find the most conserved sequence of a special pathogen
Usage
	perl conserve-seq.pl -WorkPath <path> -reference <reference genome> -isolates <isolate genomes> -algo <max or shannon> -targetLen <target length> -StorePath <output path>
Parameters
	-WorkPath	[str]	Input the analysis path
	-reference	[str]	Input the reference genome
	-isolates	[str]	Input the isolate genomes
	-algo	[str]	Choose one of the method to calculate the conservatism score [max, shannon]
	-targetLen	[int]	Input the target length
	-StorePath	[str]	Input the output path
	-h/-help	[str] print help
Auther
	Dongyan Xiong
Edit Time
	2022.07.16 21:11 0.0.3
=cut

use strict;
use warnings;
use Bio::SeqIO;
use Getopt::Long;

my $perl = "/home/dell/miniconda3/envs/Perl/bin/perl";
my $conserveregion = "/var/www/other_scripts/bio-pipelines/Conserve-sequence/Functions/conserveregion.pl";
my $findconserve = "/var/www/other_scripts/bio-pipelines/Conserve-sequence/Functions/findconserve.pl";

my $analysis_path = "";
my ($isolates, $reference, $algorithm, $targetLen, $help) = ("", "", "", 200, "");
my $storage_path = "";
GetOptions(
	'WorkPath=s' => \$analysis_path,
	'isolates=s' => \$isolates,
	'reference=s' => \$reference,
	'algo=s' => \$algorithm,
	'targetLen=i' => \$targetLen,
	'StorePath=s' => \$storage_path,
	'h|help:s' => \$help,
);
die `pod2text $0` if((!$analysis_path) or (!$isolates) or (!$reference) or (!$algorithm) or (!$targetLen) or (!$storage_path) or ($help));

my @shell;
push @shell, "#!/usr/bin/bash\n";
my $refseq = Bio::SeqIO -> new(
	-file => $analysis_path.$reference,
	-format => 'fasta',
);
my $refgenome = $refseq -> next_seq;
my $sequence = $refgenome -> seq;
$sequence = ">Ref\n".$sequence."\n";
unlink $analysis_path.$reference;
open(OUT, ">", $analysis_path.$reference);
print OUT $sequence;
close OUT;
push @shell, $perl." ".$conserveregion." -WorkPath ".$analysis_path." -reference ".$analysis_path.$reference." -isolates ".$analysis_path.$isolates." -algo ".$algorithm."\n";
push @shell, "chmod -R 777 ".$analysis_path."\n";
push @shell, $perl." ".$findconserve." -WorkPath ".$analysis_path." -basescore ".$analysis_path."base_score.tsv -stats ".$analysis_path."stat.tsv -targetLen ".$targetLen."\n";
push @shell, "chmod -R 777 ".$analysis_path."\n";
push @shell, "zip -qjr ".$analysis_path."MostConserve_info.zip ".$analysis_path."Conserved_region_info.tsv ".$analysis_path."conserved-sequence.fasta\n";
push @shell, "chmod -R 777 ".$analysis_path."\n";
push @shell, "mkdir ".$storage_path."\n";
push @shell, "chmod -R 777 ".$storage_path."\n";
push @shell, "mv ".$analysis_path."MostConserve_info.zip ".$storage_path."\n";
push @shell, "chmod -R 777 ".$storage_path."\n";
push @shell, "rm -rf ".$analysis_path."\n";
open(OUT, ">", $analysis_path.".../shell.sh");
print OUT @shell;
close OUT;
